### ISSUE TITLE
ILRepack Newtonsoft.Net

### ISSUE THREAD

**Issue by Pliner** [PlayFab/consuldotnet#28]
Hi guys!

Thank you for the library :+1: 

But we have one small trouble using your library - Newtonsoft.Json dependency. Our code base depends on 6.0.8 version and it's very difficult to update it to latest.

What do you think about ilmerging this dependency?

**Comment by highlyunavailable** [comment]
@Pliner Thanks for the suggestion. I totally get that, and 6.0.8 works fine with this code - I've tested it with that version as well. The fastest way to get this working would be to try [assemblyBinding](https://msdn.microsoft.com/en-us/library/twy1dw1e%28v=vs.110%29.aspx) the 7.0.1 to 6.0.8 I just tried this:

```
      <dependentAssembly>
        <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" culture="neutral" />
        <bindingRedirect oldVersion="7.0.0.0" newVersion="6.0.0.0" />
      </dependentAssembly>
```

And it compiled and worked just fine.

Alternatively, you could clone this repo, then change the dependency in packages.json to be `Newtonsoft.Json 6.0.8` and build your own Consul.dll but then you lose the nuget updates.

I know that @kharabasz wanted to move the Json.NET dependency into a second assembly in #22, so there could be a 608 adapter and a 701 adapter. Ideally I could express it as "Newtonsoft.Json >= 6.0.8" but it doesn't seem like .NET can actually do that without recompilation. It's such a weird requirement, though, because honestly, there are only about 4 lines of code that actually use `Newtonsoft.Json` in this library.

**Comment by highlyunavailable** [comment]
So to answer your question, I have no problem ILMerging the dependency, but I don't think it needs to be done unless binding redirects won't work for you. Let me know if you still need this.

**Comment by kharabasz** [comment]
@highlyunavailable this is a problem with Newtonsoft - I think the library is not versioned correctly. RestSharp used to use it internally and ended up dropping it in favor of allowing it to be injected for the same reasons.

**Comment by highlyunavailable** [comment]
Well, I'm happy to use something else - the actual needs are fairly lightweight. I'm going to have to think about this for a bit, but @Pliner, the binding redirect should help I think.

The only real requirement is I also need to be able to run the (de)serializer on DNXCore  (#26) so that removes a lot of the `System` libraries for serialization because they got removed in .NET 5.0.

**Comment by Pliner** [comment]
@highlyunavailable thank you for reply!

Yeap, binding redirect works fine, but there is one more problem - every time json.net is updated to new version in consul.net, we should update a lot of solutions :(

The proposition was to hide the dependency, because Consul.Net is the "core" library. 

For instance, https://github.com/facebook-csharp-sdk/simple-json could be used as lightweight alternative.

@kharabasz It could be solved in this way,the usability of the library will suffer a bit.

PS In EasyNetQ we internalize the dependency :)

**Comment by highlyunavailable** [comment]
@Pliner You're right, though I believe json.net only updates their version number every major version number, and they COULD change an API in say, 8.0 that could break a binding redirect. It's not a perfect fix but as I said it should get you working _until we have a more permanent solution_.

If you'd like to submit a PR changing the build process to do what you propose, I'd be happy to merge it, otherwise you'll have to wait until I figure out how to do this. :smile:

**Comment by Pliner** [comment]
@highlyunavailable Thank you! Will make PR as soon as possible!

**Comment by grounded042** [comment]
I like the idea of using ILRepack for this - I have a working solution and will submit a PR if I can get everything working. The one question is, how do we want to verify that this works? I can pull it into some projects that have a newtonsoft.json version < 7.0.0 and see if that works, but if others could do the same (maybe @Pliner), that would be great.

**Comment by highlyunavailable** [comment]
IMO, this package should use the latest JSON.NET and then repack could take care of all older versions. The other question is how this will interact with Mono/DNXCore, since it seems a bunch of people use this in those environments.

**Comment by grounded042** [comment]
With repack you wouldn't need to worry about older versions, it would pack the latest (or whatever you have at that moment) into Consul.dll. Also, ILRepack is done using an MSBuild.Task and I would assume that since MSBuild is being ported over to DNXCore, it should port over easily. Also, ILRpack does work with Mongo.

**Comment by highlyunavailable** [comment]
Yeah, that's what I meant re: older versions (In that I'd make consul.net use 8.x and then repack that into consul.dll).

MSBuild is _not_ part of DNXCore AFAIK - it's a very different system that doesn't use MSBuild at all. Take a look at  https://github.com/PlayFab/consuldotnet/tree/dnxcore50 specifically https://github.com/PlayFab/consuldotnet/blob/dnxcore50/Consul/project.json - that's the project file, and it doesn't look like there are easily pluggable build steps. Maybe something to do with the xproj: https://github.com/PlayFab/consuldotnet/blob/dnxcore50/Consul/Consul.xproj but that's only part of Visual Studio, I don't think the DNX build pipeline invokes it.

If it works with Mono though, that coves 95% of the "not VS" use case so I'm fine with that. I don't know of any users right now who do use this with DNXCore, just Mono but I'd like to mainstream the DNX support in the future at some point.

**Comment by grounded042** [comment]
Ah, ok, I must have been thinking of something else for MSBuild. I will convert my PR to not use the non-msbuild version of ILRepack which supports mono.

**Comment by highlyunavailable** [comment]
I think it's fine for now - I'll try it out in Mono in the next few days just for a giggle and see what happens.

**Comment by grounded042** [comment]
ok, cool - it should be all good for merge now as the tests passed on AppVeyor

**Comment by highlyunavailable** [comment]
This has been merged and released as 0.6.4.2: https://www.nuget.org/packages/Consul/0.6.4.2

> Dependencies
> 
> This package has no dependencies."

Is pretty cool to see. Thanks again @grounded042
